### PR TITLE
Add type definitions for d3-require

### PR DESCRIPTION
**Pull request by kindy**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/d3-require/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * @param name Any package name optionally followed by the 'at sign' and a semver range
+ * @returns A promise to the loaded module
+ */
+export const require: require;
+
+/**
+ * Returns a new `require` which loads modules from the specified `resolver`.
+ * Like `requireFrom(name => ('https://unpkg.com/'+name))`
+ */
+export const requireFrom: requireFrom;
```
Comment (rbuckton) [inline types/d3-require/index.d.ts]: It looks like `d3-require` is a UMD module and can be used in scripts via the `d3` global. I would recommend adding the following at the bottom:

```ts
export as namespace d3
```

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: types/d3-require/index.d.ts (hunk 1 above)
Comment (kindy) [reply types/d3-require/index.d.ts]: will add that

Comment (kindy) [reply types/d3-require/index.d.ts]: Should I add some tests for this namespace?
is this correct? Thanks. @rbuckton 

```diff
diff --git a/types/d3-require/d3-require-tests.ts b/types/d3-require/d3-require-tests.ts
index f77d0e2250..78e79158a9 100644
--- a/types/d3-require/d3-require-tests.ts
+++ b/types/d3-require/d3-require-tests.ts
@@ -1,4 +1,5 @@
 import { require as d3Require, requireFrom } from 'd3-require';
+import d3 = require('d3-require');
 
 d3Require('abc').then(x => null);
 
@@ -6,3 +7,5 @@ d3Require.alias({a: 'a'})('a', 'b').then(x => null);
 d3Require.resolve('b').then(x => x.length);
 
 requireFrom(async a => ('/~/' + a))('a', 'b').then(x => null);
+
+d3.require('abc@x');
```

Comment (rbuckton) [reply types/d3-require/index.d.ts]: My apologies. After further review we probably should *not* use `export as namespace d3;`. This package is intended to be used alongside the `d3` package which *also* has an `export as namespace d3`, and using both together would result in a conflict.

I'm not sure how best to resolve that conflict at the moment, so for the time being I would remove the `export as namespace d3` line and I'll go ahead and merge.

Comment (kindy) [reply types/d3-require/index.d.ts]: namespace removed.

File: types/d3-require/tslint.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+    "extends": "dtslint/dt.json",
+    "rules": {
+        "unified-signatures": false,
+        "no-unnecessary-generics": false
```
Comment (DanielRosenwasser) [inline types/d3-require/tslint.json]: Why do you need these?

Comment (kindy) [reply types/d3-require/tslint.json]: @DanielRosenwasser  this **tslint.json** file is copy from other module, should I fix it?

Comment (DanielRosenwasser) [reply types/d3-require/tslint.json]: Yes please

**Comment by orta** [comment]
I've given this a look over, @kindy - does this look good to go now?

**Pull request merged**